### PR TITLE
Make warnings in Java code fatal

### PR DESCRIPTION
**Pull request by raboof**

### CODE REVIEW

File: akka-persistence-typed/src/test/java/jdocs/akka/persistence/typed/WebStoreCustomerFSM.java [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+
+    public List<Item> getItems() {
+      return items;
+    }
+  }
+
+  public enum ShoppingCardDiscarded implements ReportEvent {
+    INSTANCE
+  }
+}
```
Comment (helena) [inline akka-persistence-typed/src/test/java/jdocs/akka/persistence/typed/WebStoreCustomerFSM.java]: How is this related to fatal warnings?

File: project/AkkaDisciplinePlugin.scala [hunk 2]
```diff
@@ -18,6 +18,18 @@ object AkkaDisciplinePlugin extends AutoPlugin with ScalafixSupport {
 
   // We allow warnings in docs to get the 'snippets' right
   val nonFatalWarningsFor = Set("akka-docs")
+  val nonFatalJavaWarningsFor = Set(
+    // for sun.misc.Unsafe and AbstractScheduler
+    "akka-actor",
+    // use of deprecated akka.protobuf.GeneratedMessage
+    "akka-protobuf",
```
Comment (helena) [inline project/AkkaDisciplinePlugin.scala]: If this fixes what I think it does, 👍

**Review by helena — APPROVED** [review]
LGTM

File: akka-persistence-typed/src/test/java/jdocs/akka/persistence/typed/WebStoreCustomerFSM.java (hunk 1 above)
Comment (raboof) [reply akka-persistence-typed/src/test/java/jdocs/akka/persistence/typed/WebStoreCustomerFSM.java]: This new file contains the model needed for the `PersistentFsmToTypedMigrationCompileOnlyTest`.

Previously that re-used the model from `akka.persistence.fsm.AbstractPersistentFSMTest`, but it can no longer import that because `akka.persistence.fsm` is deprecated.

File: project/Dependencies.scala [hunk 3]
```diff
@@ -37,12 +37,7 @@ object Dependencies {
     crossScalaVersions := Seq(scala212Version, scala213Version),
     scalaVersion := System.getProperty("akka.build.scalaVersion", crossScalaVersions.value.head),
     scalaCheckVersion := sys.props.get("akka.build.scalaCheckVersion").getOrElse("1.14.2"),
-    scalaTestVersion := {
-      CrossVersion.partialVersion(scalaVersion.value) match {
-        case Some((2, n)) if n >= 13 => "3.0.8"
-        case _                       => "3.0.7"
-      }
-    },
+    scalaTestVersion := "3.0.8",
```
Comment (chbatey) [inline project/Dependencies.scala]: why were we sticking to 3.0.7? DId they make an incompatible change?

**Review by chbatey — APPROVED** [review]
LGTM, with one question about the upgrade of scalatest

File: project/Dependencies.scala (hunk 3 above)
Comment (raboof) [reply project/Dependencies.scala]: Hmm, didn't remember there were so many scalatest changes in this PR. Perhaps we should just merge #28473 first, then I'll rebase this one

**Comment by ennru** [comment]
This collected some dust. But at least the fixed warnings would be great to get in.

**Comment by raboof** [comment]
Rebased and updated (caught quite some deprecations in the example code, nice!), fixed #25530 in  the process.

File: akka-docs/src/test/java/jdocs/actor/ActorDocTest.java [hunk 4]
```diff
@@ -526,12 +526,14 @@ public void creatingPropsConfig() {
...
   public void creatingPropsIllegal() {
+    /*
     // #creating-props-illegal
     // This will throw an IllegalArgumentException since some runtime
     // type information of the lambda is erased.
     // Use Props.create(actorClass, Creator) instead.
     Props props = Props.create(() -> new ActorWithArgs("arg"));
     // #creating-props-illegal
+    */
   }
```
Comment (ennru) [inline akka-docs/src/test/java/jdocs/actor/ActorDocTest.java:537]: This should fail now, right?

File: akka-persistence-typed/src/test/java/jdocs/akka/persistence/typed/PersistentFsmToTypedMigrationCompileOnlyTest.java [hunk 5]
```diff
@@ -97,7 +98,7 @@ public String manifest(DomainEvent event) {
 
     @Override
     public EventSeq<DomainEvent> fromJournal(Object event, String manifest) {
-      if (event instanceof StateChangeEvent) {
+      if (event instanceof akka.persistence.fsm.PersistentFSM.StateChangeEvent) {
```
Comment (ennru) [inline akka-persistence-typed/src/test/java/jdocs/akka/persistence/typed/PersistentFsmToTypedMigrationCompileOnlyTest.java:101]: Is this consistent with the changed import?

File: akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala [hunk 6]
```diff
@@ -1514,7 +1514,6 @@ final class Flow[In, Out, Mat](delegate: scaladsl.Flow[In, Out, Mat]) extends Gr
    *
    * '''Cancels when''' downstream cancels
    */
-  @deprecated("Use recoverWithRetries instead.", "2.4.4")
```
Comment (ennru) [inline akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala:1517]: Why doesn't this apply anymore? (I guess it should have been remove on the way, but it is still deprecated, right.)

**Review by ennru — COMMENTED** [review]
Quite some things improving here.

File: akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala (hunk 6 above)
Comment (raboof) [reply akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala:1517]: Ah, this was incorrectly deprecated, so I fixed #25530 as part of this PR

File: akka-persistence-typed/src/test/java/jdocs/akka/persistence/typed/PersistentFsmToTypedMigrationCompileOnlyTest.java (hunk 5 above)
Comment (raboof) [reply akka-persistence-typed/src/test/java/jdocs/akka/persistence/typed/PersistentFsmToTypedMigrationCompileOnlyTest.java:101]: Yes: the changed import covered the `WebStoreCustomerFSM` classes from akka-persistence/src/test, while StateChangeEvent is from akka-persistence/src/main

**Review by ennru — APPROVED** [review]
LGTM.
